software/py/blood_graph.py

```diff
@@ -101,50 +285,52 @@ def mark_trace(self, trace):
-      self.pixel[col,row] = (0xff, 0x45, 0x00) # orange
-    elif trace["stall"] == "stall_depend_remote_load":
-      self.pixel[col,row] = (0xff, 0x00, 0x00) # red
-    elif trace["stall"] == "stall_depend_local_remote_load":
-      self.pixel[col,row] = (0x80, 0x00, 0x00) # maroon
-    elif trace["stall"] == "stall_fp_remote_load":
-      self.pixel[col,row] = (0x00, 0x80, 0x00) # green
-    elif trace["stall"] == "stall_fp_local_load":
-      self.pixel[col,row] = (0x00, 0x64, 0x00) # darkgreen
-    elif trace["stall"] == "stall_force_wb":
-      self.pixel[col,row] = (0x20, 0xb2, 0xaa) # lightseagreen
-    elif trace["stall"] == "stall_ifetch_wait":
-      self.pixel[col,row] = (0x00, 0x00, 0xff) # blue
-    elif trace["stall"] == "stall_icache_store":
-      self.pixel[col,row] = (0x00, 0x00, 0xff) # grey
-    elif trace["stall"] == "stall_lr_aq":
-      self.pixel[col,row] = (0xff, 0xd7, 0x00) # gold
-    elif trace["stall"] == "stall_md":
-      self.pixel[col,row] = (0x80, 0x00, 0x80) # purple
-    elif trace["stall"] == "stall_remote_req":
-      self.pixel[col,row] = (0x00, 0xff, 0xff) # cyan
-    elif trace["stall"] == "stall_local_flw":
-      self.pixel[col,row] = (0x00, 0xff, 0x00) # lime
-    elif trace["stall"] == "no_stall":
-      self.pixel[col,row] = (0xff, 0xff, 0xff) # white
-      
+    if trace["operation"] in self.stall_bubble_color.keys():
+      self.pixel[col,row] = self.stall_bubble_color[trace["operation"]]
+    elif trace["operation"] in self.instr_list:
+      self.pixel[col,row] = self.unified_instr_color
+    elif trace["operation"] in self.fp_instr_list:
+      self.pixel[col,row] = self.unified_fp_instr_color
+    elif trace["operation"] in self.unknown_list:
+      self.pixel[col,row] = self.unknown_color
+    else:
+      print ("Error: invalid operaiton in operation log: " + trace["operation"])
+      sys.exit()
+    return
 
-# main()
-if __name__ == "__main__":
+   
 
-  if len(sys.argv) != 5:
-    print("wrong number of arguments.")
-    print("python {start_time} {end_time} {timestep} vanilla.log")
+def parse_args():
+  parser = argparse.ArgumentParser(description="TODO")
+  parser.add_argument("--input", default=DEFAULT_INPUT_FILE, type=str,
+                      help="Vanilla operation log file")
+  parser.add_argument("--mode", default=DEFAULT_MODE, type=str,
+                      help="Type of bloodgraph - abstract / detailed")
+  parser.add_argument("--start", default=DEFAULT_START_TIME, type=int,
+                      help="Starting cycle of bloodgraph")
+  parser.add_argument("--end", default=DEFAULT_END_TIME, type=int,
+                      help="Ending cycle of bloodgraph")
+  parser.add_argument("--timestamp", default=DEFAULT_TIMESTAMP, type=int,
+                      help="Distance between each trace in cycles")
+  args = parser.parse_args()
+
+  if (args.start > args.end):
+    parser.error("start cycle cannot be larger than end time.")
```

Comment:
https://docs.python.org/3.6/library/argparse.html#action

Reply:
Thanks for pointing it out! I added a TimeAction to take in time as start_time@end_time and check if start time is less than end time.